emulator/src/kreator/assembler.rs

```diff
@@ -888,11 +743,60 @@ mod tests {
                 bytes.push(byte.parse::<u8>().unwrap());
             }
             let operation = components[1];
-            let assembler = Assembler::new(operation);
+            let assembler = Assembler::new(format!("{}\nEND", operation).as_str());
             assert_eq!(bytes, assembler.assemble().unwrap());
         }
     }
 
+    #[test]
+    fn org_first_address() {
+        let assembler = Assembler::new("RNC \n ORG 20H\nEND");
+        assert_eq!(vec![(1, 32)], assembler.get_origins());
+
+        let assembler = Assembler::new("RNC\nEND");
+        assert_eq!(Vec::<(u16, u16)>::new(), assembler.get_origins());
+
+        let assembler = Assembler::new("ORG 5 + 1 \nRNC\nEND");
+        assert_eq!(vec![(0, 6)], assembler.get_origins());
+    }
+
+    #[test]
+    fn multiple_orgs() {
+        let assembler = Assembler::new(
+            "ORG 1000H \n MOV A,C \n ADI 2\n JMP NEXT \n HERE:ORG 1050H \n NEXT: XRA A\nEND",
+        );
+        let jumps: Vec<(u16, u16)> = vec![(0, 0x1000), (6, 0x1050)];
+
+        assert_eq!(jumps, assembler.get_origins());
+    }
+
+    #[test]
+    fn full_program() {
+        let code = "VAR1 EQU 123\n 
+        GO: JMP $ +6\n
+        ADD C\n
+        \n
+        \n
+        IF 0+0*00O\n
+        MOV A,B\n
+        ENDIF\n
+        POP B\n
+        macr0 MACRO par\n
```

Comment:
Is this valid syntax? Sry for my ignorance, but this seems odd

Reply:
I'm not sure to what you are referring here. Are you talking about the macro declaration?

Reply:
yes, `macr0 MACRO ...` seems weird

Reply:
According to page 41 of the 8080s Programmers Manual a macro is defined  as `name MACRO list` whereas the name is part of the label field (meaning it must start with a letter, "@" or "?", and be 1-5 characters in length). The list contains the macros parameters seperated by commas, these are usually ASCII constants.

Maybe the name and name of the single parameter have been chosen in a bad way, making it less intuitive?

Reply:
no, not really. i was just being stupid.